core/src/main/java/com/linecorp/armeria/server/RoutingTrie.java

```diff
@@ -515,16 +552,29 @@ public String toString() {
          * Attaches a given {@code value} to the value list. If the list is not empty
          * the {@code value} is added, and sorted by the given {@link Comparator}.
          */
-        private void addValue(V value, @Nullable Comparator<V> comparator) {
-            if (values == null) {
-                values = new ArrayList<>();
+        private void addValue(V value, boolean highPrecedence, @Nullable Comparator<V> comparator) {
+            final List<V> values;
+            if (highPrecedence) {
+                if (valuesWithHighPrecedence == null) {
+                    valuesWithHighPrecedence = new ArrayList<>(4);
+                }
+                values = valuesWithHighPrecedence;
+            } else {
+                if (valuesWithLowPrecedence == null) {
+                    valuesWithLowPrecedence = new ArrayList<>(4);
+                }
+                values = valuesWithLowPrecedence;
             }
+
             values.add(value);
 
             // Sort the values using the given comparator.
             if (comparator != null && values.size() > 1) {
                 values.sort(comparator);
             }
+
+            // Invalidate the copy of concat(nonRedirectValues, redirectValues).
+            allValues = null;
```

Comment:
I think this is thread-safe only because we assume reads of `allValues` will only ever happen on a built router. Doing this explicitly by separating `NodeBuilder` and `Node` would be clearer IMO and maybe somewhat better performance.

Reply:
Agreed. `RoutingTrie` is now immutable.